google/resource_storage_notification_test.go

```diff
@@ -0,0 +1,250 @@
+					resource.TestCheckResourceAttr(
+						"google_storage_notification.notification", "topic", topic),
+					resource.TestCheckResourceAttr(
+						"google_storage_notification.notification", "payload_format", payload),
+					testAccCheckStorageNotificationCheckEventType(
+						&notification, []string{eventType1, eventType2}),
+					testAccCheckStorageNotificationCheckAttributes(
+						&notification, "new-attribute", "new-attribute-value"),
+				),
+			},
+			resource.TestStep{
+				ResourceName:      "google_storage_notification.notification",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func testAccGoogleStorageNotificationDestroy(s *terraform.State) error {
+	config := testAccProvider.Meta().(*Config)
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "google_storage_notification" {
+			continue
+		}
+
+		bucket := rs.Primary.Attributes["bucket"]
+
+		_, notificationID := resourceStorageNotificationParseID(rs.Primary.ID)
+
+		_, err := config.clientStorage.Notifications.Get(bucket, notificationID).Do()
+		if err == nil {
+			return fmt.Errorf("Notification configuration still exists")
+		}
+	}
+
+	return nil
+}
+
+func testAccCheckStorageNotificationExists(bucket, resource string, notification *storage.Notification) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resource]
+		if !ok {
+			return fmt.Errorf("Not found: %s", resource)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No ID is set")
+		}
+
+		config := testAccProvider.Meta().(*Config)
+
+		_, notificationID := resourceStorageNotificationParseID(rs.Primary.ID)
+
+		found, err := config.clientStorage.Notifications.Get(bucket, notificationID).Do()
+		if err != nil {
+			return err
+		}
+
+		if found.Id != notificationID {
+			return fmt.Errorf("Storage notification configuration not found")
+		}
+
+		*notification = *found
+
+		return nil
+	}
+}
+
+func testAccCheckStorageNotificationCheckEventType(notification *storage.Notification, eventTypes []string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+
+		if !reflect.DeepEqual(notification.EventTypes, eventTypes) {
+			return fmt.Errorf("Target event types are incorrect. Expected %s, got %s", eventTypes, notification.EventTypes)
+		}
+		return nil
+	}
+}
+
+func testAccCheckStorageNotificationCheckAttributes(notification *storage.Notification, key, value string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		val, ok := notification.CustomAttributes[key]
+		if !ok {
+			return fmt.Errorf("Custom attribute with key %s not found", key)
+		}
+
+		if val != value {
+			return fmt.Errorf("Custom attribute value did not match for key %s: expected %s but found %s", key, value, val)
+		}
+		return nil
+	}
+}
+
+func testGoogleStorageNotificationBasic(bucketName, topicName, topic string) string {
+	return fmt.Sprintf(`
+		resource "google_storage_bucket" "bucket" {
```

Comment:
nit: we try to have these all the way on the left (mostly a personal preference thing, it just makes it a bit easier to copy/paste into our own configs later, feel free to ignore)